cloud/scope/machine.go

```diff
@@ -299,14 +264,97 @@ func (m *MachineScope) GetOrCreateMachine(ctx context.Context) (*core.Instance,
 	launchDetails.PreemptibleInstanceConfig = m.getPreemptibleInstanceConfig()
 	launchDetails.PlatformConfig = m.getPlatformConfig()
 	launchDetails.LaunchVolumeAttachments = m.getLaunchVolumeAttachments()
-	req := core.LaunchInstanceRequest{LaunchInstanceDetails: launchDetails,
-		OpcRetryToken: ociutil.GetOPCRetryToken(string(m.OCIMachine.UID))}
-	resp, err := m.ComputeClient.LaunchInstance(ctx, req)
+	// Build the list of fault domains to attempt launch in
+	faultDomains := m.buildFaultDomainLaunchList(availabilityDomain, faultDomain, retry)
+	return m.launchInstanceWithFaultDomainRetry(ctx, launchDetails, faultDomains)
+}
+
+func (m *MachineScope) launchInstanceWithFaultDomainRetry(ctx context.Context, baseDetails core.LaunchInstanceDetails, faultDomains []string) (*core.Instance, error) {
+	if len(faultDomains) == 0 {
+		faultDomains = []string{ociutil.DerefString(baseDetails.FaultDomain)}
+	}
+	opcRetryToken := ociutil.GetOPCRetryToken(string(m.OCIMachine.UID))
+	var lastErr error
+	for idx, fd := range faultDomains {
+		details := baseDetails
+		if fd != "" {
+			details.FaultDomain = common.String(fd)
+		}
+		resp, err := m.ComputeClient.LaunchInstance(ctx, core.LaunchInstanceRequest{
+			LaunchInstanceDetails: details,
+			OpcRetryToken:         opcRetryToken,
+		})
+		if err == nil {
+			return &resp.Instance, nil
+		}
+		lastAttempt := idx == len(faultDomains)-1
+		if !ociutil.IsOutOfHostCapacity(err) || lastAttempt {
+			return nil, err
+		}
+		lastErr = err
+		m.Logger.Info("Fault domain has run out of host capacity, retrying in a different domain", "faultDomain", fd)
+	}
+	return nil, lastErr
+}
+
+func (m *MachineScope) buildFaultDomainLaunchList(availabilityDomain, initialFaultDomain string, retry bool) []string {
```

Comment:
Maybe rename to something more clear like `buildFaultDomainRetryList` or something like that. Since this is being used for retry logic.